rs/nns/governance/src/governance/voting_power_snapshots.rs

```diff
@@ -0,0 +1,159 @@
+#![allow(unused)]
+use crate::{
+    governance::LOG_PREFIX,
+    neuron_store::voting_power::VotingPowerSnapshot,
+    pb::v1::{Ballot, VotingPowerMap, VotingPowerTotal},
+};
+
+use ic_stable_structures::{
+    memory_manager::VirtualMemory, storable::Bound, DefaultMemoryImpl, StableBTreeMap, Storable,
+};
+use prost::Message;
+use std::{borrow::Cow, collections::HashMap};
+
+const MAX_VOTING_POWER_SNAPSHOTS: u64 = 7;
+const MULTIPLIER_THRESHOLD_FOR_VOTING_POWER_SPIKE: u64 = 2;
+
+type VM = VirtualMemory<DefaultMemoryImpl>;
+
+pub(crate) struct VotingPowerSnapshots {
+    voting_power_maps: StableBTreeMap<u64, VotingPowerMap, VM>,
+    voting_power_totals: StableBTreeMap<u64, VotingPowerTotal, VM>,
+}
+
+fn insert_and_truncate<Value: Storable>(
+    map: &mut StableBTreeMap<u64, Value, VM>,
+    timestamp_seconds: u64,
+    value: Value,
+) {
+    let existing_value = map.insert(timestamp_seconds, value);
+    if let Some(existing_value) = existing_value {
+        ic_cdk::eprintln!(
+            "{}Somehow the voting power snapshot is taken multiple times at \
+            the same timestamp {}",
+            LOG_PREFIX,
+            timestamp_seconds,
+        );
+    }
+    while map.len() > MAX_VOTING_POWER_SNAPSHOTS {
+        let (first_key, _) = map
+            .first_key_value()
+            .expect("No first key value even though the length is checked right before.");
+        map.remove(&first_key);
+    }
+}
+
+impl VotingPowerSnapshots {
+    pub fn new(maps_memory: VM, totals_memory: VM) -> Self {
```

Comment:
> totals_memory

Isn't this going to be quite small? Isn't it a bit overkill to store it in stable memory?